Realm/RLMRealmUtil.mm

```diff
@@ -77,227 +85,35 @@ void RLMInstallUncaughtExceptionHandler() {
     }
-    return self;
-}
-
-- (void)listen {
-
-    // Create the runloop source
-    CFRunLoopSourceContext ctx{};
-    ctx.info = (__bridge void *)self;
-    ctx.perform = [](void *info) {
-        RLMNotifier *notifier = (__bridge RLMNotifier *)info;
-        [notifier->_realm handleExternalCommit];
-    };
-
-    CFRunLoopSourceRef signal = CFRunLoopSourceCreate(kCFAllocatorDefault, 0, &ctx);
-    CFRunLoopAddSource(_runLoop, signal, kCFRunLoopDefaultMode);
-
-    // Set up the kqueue
-    // EVFILT_READ indicates that we care about data being available to read
-    // on the given file descriptor.
-    // EV_CLEAR makes it wait for the amount of data available to be read to
-    // change rather than just returning when there is any data to read.
-    struct kevent ke[2];
-    EV_SET(&ke[0], _notifyFd, EVFILT_READ, EV_ADD | EV_CLEAR, 0, 0, 0);
-    EV_SET(&ke[1], _shutdownReadFd, EVFILT_READ, EV_ADD | EV_CLEAR, 0, 0, 0);
-    int ret = kevent(_kq, ke, 2, nullptr, 0, nullptr);
-    assert(ret == 0);
-
-    while (true) {
-        struct kevent event;
-        // Wait for data to become on either fd
-        // Return code is number of bytes available or -1 on error
-        ret = kevent(_kq, nullptr, 0, &event, 1, nullptr);
-        assert(ret >= 0);
-        if (ret == 0) {
-            // Spurious wakeup; just wait again
-            continue;
-        }
 
-        // Check which file descriptor had activity: if it's the shutdown
-        // pipe, then someone called -stop; otherwise it's the named pipe
-        // and someone committed a write transaction
-        if (event.ident == (uint32_t)_shutdownReadFd) {
-            CFRunLoopSourceInvalidate(signal);
-            CFRelease(signal);
-            CFRelease(_runLoop);
-            return;
-        }
-        assert(event.ident == (uint32_t)_notifyFd);
+    void will_change(std::vector<ObserverState> const& observed, std::vector<void*> const& invalidated) override {
+        RLMWillChange(observed, invalidated);
+    }
 
-        CFRunLoopSourceSignal(signal);
-        // Signalling the source makes it run the next time the runloop gets
-        // to it, but doesn't make the runloop start if it's currently idle
-        // waiting for events
-        CFRunLoopWakeUp(_runLoop);
+    void did_change(std::vector<ObserverState> const& observed, std::vector<void*> const& invalidated) override {
+        RLMDidChange(observed, invalidated);
+        [_realm sendNotifications:RLMRealmDidChangeNotification];
     }
-}
 
-- (void)stop {
-    notifyFd(_shutdownWriteFd);
-}
+private:
+    // This is owned by the realm, so it needs to not retain the realm
+    __unsafe_unretained RLMRealm *const _realm;
+};
+} // anonymous namespace
+
 
-- (void)notifyOtherRealms {
-    notifyFd(_notifyFd);
+std::unique_ptr<realm::RealmDelegate> RLMCreateRealmDelegate(RLMRealm *realm) {
+    return std::unique_ptr<realm::RealmDelegate>(new RLMNotificationHelper(realm));
```

Comment:
The type of the object being constructed and the type of the pointer being returned don't match (since the whole point is to avoid having to expose the definition of the concrete type and thus drag in extra headers), and `unique_ptr<Derived>` doesn't implicitly convert to `unique_ptr<Base>`.